src/protocol/aggregation/mod.rs

```diff
@@ -52,139 +58,229 @@ where
+    f: H,
+) -> Result<Vec<BitDecomposed<S>>, Error>
 where
+    F: PrimeField,
+    C: UpgradedContext<F, Share = S>,
+    S: LinearSecretSharing<F> + BasicProtocols<C, F> + Serializable + 'static,
+    H: Fn(&SparseAggregateInputRow<CV, BK>, u32) -> Replicated<Gf2>,
     CV: GaloisField,
     BK: GaloisField,
 {
-    input_rows
+    let num_records = input_rows.len();
+    let gf2_bits = input_rows
         .iter()
-        .map(|row| {
-            BitDecomposed::decompose(CV::BITS, |i| {
-                Replicated::new(
-                    Gf2::truncate_from(row.contribution_value.left()[i]),
-                    Gf2::truncate_from(row.contribution_value.right()[i]),
-                )
-            })
-        })
-        .collect::<Vec<_>>()
-}
+        .map(|row| BitDecomposed::decompose(num_bits, |i| f(row, i)))
+        .collect::<Vec<_>>();
 
-fn get_gf2_breakdown_key_bits<CV, BK>(
-    input_rows: &[SparseAggregateInputRow<CV, BK>],
-) -> Vec<BitDecomposed<Replicated<Gf2>>>
-where
-    CV: GaloisField,
-    BK: GaloisField,
-{
-    input_rows
-        .iter()
-        .map(|row| {
-            BitDecomposed::decompose(BK::BITS, |i| {
-                Replicated::new(
-                    Gf2::truncate_from(row.breakdown_key.left()[i]),
-                    Gf2::truncate_from(row.breakdown_key.right()[i]),
-                )
-            })
-        })
-        .collect::<Vec<_>>()
+    convert_bits(
+        ctx.narrow(&Step::ConvertValueBits)
+            .set_total_records(num_records),
+        stream_iter(gf2_bits),
+        0..num_bits,
+    )
+    .try_collect::<Vec<_>>()
+    .await
 }
 
 #[cfg(all(test, unit_test))]
 mod tests {
-    use super::aggregate_values;
+    use super::sparse_aggregate;
     use crate::{
-        ff::{Fp32BitPrime, Gf2},
-        protocol::context::{UpgradableContext, Validator},
-        secret_sharing::BitDecomposed,
+        ff::{Field, Fp32BitPrime, GaloisField, Gf3Bit, Gf8Bit},
+        protocol::aggregation::SparseAggregateInputRow,
+        secret_sharing::replicated::semi_honest::AdditiveShare as Replicated,
         test_fixture::{Reconstruct, Runner, TestWorld},
     };
 
+    fn create_input_vec<T, U>(
+        input: &[(Replicated<T>, Replicated<U>)],
+    ) -> Vec<SparseAggregateInputRow<T, U>>
+    where
+        T: GaloisField,
+        U: GaloisField,
+    {
+        input
+            .iter()
+            .map(|x| SparseAggregateInputRow {
+                contribution_value: x.0.clone(),
+                breakdown_key: x.1.clone(),
+            })
+            .collect::<Vec<_>>()
+    }
+
     #[tokio::test]
     pub async fn aggregate() {
-        const CONTRIBUTION_BITS: u32 = 8;
-        const EXPECTED: u128 = 36;
+        type CV = Gf8Bit;
+        type BK = Gf3Bit;
 
-        const INPUT: &[u32] = &[0, 0, 18, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 10, 0, 0, 6, 0];
+        const EXPECTED: &[u128] = &[28, 0, 0, 6, 1, 0, 0, 8];
+
+        const INPUT: &[(u32, u32)] = &[
```

Comment:
I had trouble reading the ordering here.  I read this as a k-v map, where it is a v-k map instead.

Reply:
I also got caught up on the same thing.